### ISSUE TITLE
Don't emit an implicit Object superclass for interface types. [phab:D1004]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1340]
_Created by **benjyw** at 2016-08-10 23:10:11_: 

It is true that all interfaces are subtypes of Object
(see http://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.10.2),
but an interface does not extend Object and is therefore not
a nominal subtype, as implied by the EXTENDS edge.

(The fact that Kythe emits that edge was carelessness on my 
part in 2f10c61be6.  The intention was only to emit for classes,
so that the behavior would be uniform whether the "extends Object"
was implicit or explicit. This is not pertinent to interfaces.)

**Comment by kythe** [comment]
_**zarko** wrote at 2016-08-11 19:32:45_:

It's not clear to me that this is the right thing to do; for example, this compiles:

```
public class If {
  public interface It { }
  public static class Iti implements It { }
  public static void fn(Object o) { }
  public static void main(String[] args) {
    It it = new Iti();
    fn(it);
  }
}
```
Non-primitive types in Java are always (nominal) subtypes of some other type, which (at the top of the lattice) is Object. Why should interface types be excluded?

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-08-11 19:54:07_:

TBH I could go either way here, this change was the easiest way to have the conversation...

Here are the arguments in favor:

- According to the java spec, an interface is a subtype of Object, but it does *not* "extend" Object.  True, you can call Object's methods through an interface-typed reference. But this is implemented by the compiler synthesizing those methods directly onto every interface, not by having interface extend Object. So this is more like duck-typing than nominal subtyping.

- If you examine the .class file contents, a class is always specified to extend Object, but an interface never is.

I think the latter argument is strong, the former is a matter of semantics.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-08-11 20:20:04_:

The `extends` edge doesn't necessarily mean the same thing as Java's `extends` (or `implements`) keyword; we generally prefer arguments grounded in language semantics (or statics), not in implementation details (like what one implementation of javac might compile to class files).

My understanding of the language definition is that you can use a reference of interface type `I` where a reference of `Object` type is expected because `I` is a subtype of `Object`. Is this incorrect? If so, is there some special coercion that happens when (e.g.) a reference to an object of type `I` is assigned to a field that stores references to objects of type `Object`?

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-08-12 19:04:32_:

>>! In (earlier comment), zarko wrote:
> The `extends` edge doesn't necessarily mean the same thing as Java's `extends` (or `implements`) keyword; we generally prefer arguments grounded in language semantics (or statics), not in implementation details (like what one implementation of javac might compile to class files).
> 
> My understanding of the language definition is that you can use a reference of interface type `I` where a reference of `Object` type is expected because `I` is a subtype of `Object`. Is this incorrect? If so, is there some special coercion that happens when (e.g.) a reference to an object of type `I` is assigned to a field that stores references to objects of type `Object`?

It is correct that `I` is a subtype of `Object` (see http://docs.oracle.com/javase/specs/jls/se7/html/jls-4.html#jls-4.10.2), but I don't know that I would say that it's a nominal subtype. 

http://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.4.1.2 states:

```interfaces do not inherit from Object, but rather implicitly declare many of the same methods as Object```

And http://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.2 elaborates.

So you can always assign something referenced by `I` to an `Object` reference, but as far as I can tell this is due to special casing.

That said, as I mentioned, I could go either way on this.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-08-15 17:06:40_:

Aha, you've convinced me with 9.4.1.2, especially when we include a little more text:

> So, there is no common ancestor for the toString declared in Object and the toString declared in an interface. At best, if both were candidates for inheritance by a class, they would conflict. Working around this problem would require awkward commingling of the class and interface inheritance trees.

We have to defer to the JLS if it insists that interface inheritance is disjoint from class inheritance. (I will limit my editorializing to saying that this is gross and weird, especially all the special case rules in 9.2.)

Thanks for the diff (and for hunting around in the JLS)!

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-08-15 17:06:40_

**Comment by kythe** [comment]
_**zarko** closed via commit f1ff5912b886f8b4ae69df1db4b1d15daf5bc868 at 2016-08-15 17:08:36_

**Comment by kythe** [comment]
_**benjyw** wrote at 2016-08-16 01:49:21_:

Thanks for the review!
